app/models/transformation_mapping.rb

```diff
@@ -1,9 +1,115 @@
 class TransformationMapping < ApplicationRecord
+  VM_CONFLICT = N_("Conflict").freeze
+  VM_NON_EXIST = N_("VM does not exist").freeze
+  VM_VALID = N_("OK").freeze
+
   has_many :transformation_mapping_items, :dependent => :destroy
 
   validates :name, :presence => true, :uniqueness => true
 
   def destination(source)
     transformation_mapping_items.find_by(:source => source).try(:destination)
   end
+
+  # vm_list: collection of hashes, each descriping a VM.
+  def validate_vms(vm_list = nil)
+    vm_list.present? ? identify_vms(vm_list) : select_vms
+  end
+
+  private
+
+  def select_vms
+    valid_list = []
+
+    transformation_mapping_items.where(:source_type => EmsCluster).collect(&:source).each do |cluster|
+      cluster.vms.each do |vm|
+        reason = validate_vm(vm)
+        valid_list << describe_vm(vm, reason) if reason == VM_VALID
+      end
+    end
+
+    {"valid_vms" => valid_list}
+  end
+
+  def identify_vms(vm_list)
+    valid_list = []
+    invalid_list = []
+    conflict_list = []
+
+    vm_list.each do |row|
+      vm_name = row['name']
+
+      query = Vm.where(:name => vm_name)
+      query = query.where(:uid_ems => row['uid_ems']) if row['uid_ems'].present?
+      query = query.joins(:host).where(:hosts => {:name => row['host']}) if row['host'].present?
+      query = query.joins(:ext_management_system).where(:ext_management_systems => {:name => row['provider']}) if row['provider'].present?
+
+      vms = query.to_a
+      invalid_list << describe_non_vm(vm_name) if vms.empty?
+
+      if vms.size == 1
+        reason = validate_vm(vms.first)
+        (reason == VM_VALID ? valid_list : invalid_list) << describe_vm(vms.first, reason)
+      else
+        vms.each { |vm| conflict_list << describe_vm(vm, VM_CONFLICT) }
+      end
+    end
+
+    {
+      "valid_vms"    => valid_list,
+      "invalid_vms"  => invalid_list,
+      "conflict_vms" => conflict_list
+    }
+  end
+
+  def describe_non_vm(vm_name)
+    {
+      "name"   => vm_name,
+      "reason" => VM_NON_EXIST
+    }
+  end
+
+  def describe_vm(vm, reason)
+    # formate to be finalized
+    {
+      "name"           => vm.name,
+      "cluster"        => vm.ems_cluster.name,
+      "path"           => "#{vm.ext_management_system.name}/#{vm.parent_blue_folder_path(:exclude_non_display_folders => true)}",
+      "allocated_size" => vm.allocated_disk_storage,
+      "id"             => vm.id,
+      "reason"         => reason
+    }
+  end
+
+  def validate_vm(vm)
```

Comment:
We can change to `def validate_vm(vm, quick)` with `quick` to be a boolean. When it is `true` for the use by `select_vms`, it should quickly exit once a problem is detected. When it is `false` for the use by `identify_vms` it should go through all the validations to collect a complete description of all problems.